lib/Transforms/IndexSwitchToIf.cpp

```diff
@@ -0,0 +1,118 @@
+//===- IndexSwitchToIf.cpp - Index switch to if-else pass ---*-C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the SCF IndexSwitch to If-Else pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Transforms/Passes.h"
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+#define GEN_PASS_DEF_INDEXSWITCHTOIF
+#include "circt/Transforms/Passes.h.inc"
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+struct SwitchToIfConversion : public OpConversionPattern<scf::IndexSwitchOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(scf::IndexSwitchOp switchOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = switchOp.getLoc();
+
+    Region &defaultRegion = switchOp.getDefaultRegion();
+
+    Value finalResult;
+    scf::IfOp prevIfOp = nullptr;
+
+    rewriter.setInsertionPointAfter(switchOp);
+    auto switchCases = switchOp.getCases();
+    for (size_t i = 0; i < switchCases.size(); i++) {
+      auto caseValueInt = switchCases[i];
+      if (prevIfOp)
+        rewriter.setInsertionPointToStart(&prevIfOp.getElseRegion().front());
+
+      Value caseValue =
+          rewriter.create<arith::ConstantIndexOp>(loc, caseValueInt);
+      Value cond = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::eq, switchOp.getOperand(), caseValue);
+
+      auto ifOp = rewriter.create<scf::IfOp>(loc, switchOp.getResultTypes(),
+                                             cond, /*hasElseRegion=*/true);
+
+      Region &caseRegion = switchOp.getCaseRegions()[i];
+      IRMapping mapping;
+      Block &emptyThenBlock = ifOp.getThenRegion().front();
+      emptyThenBlock.erase();
+      caseRegion.cloneInto(&ifOp.getThenRegion(), mapping);
+
+      if (i == switchCases.size() - 1) {
```

Comment:
Sorry one last small nit: be careful subtracting unsigned numbers, sometimes you don't want the modular arithmetic.
```suggestion
      if (i + 1 == switchCases.size()) {
```

Reply:
thanks for pointing it out, makes sense!